src/cookbooks/bigquery/optimization.md

```diff
@@ -1,28 +1,165 @@
+- Use approximate algorithms: e.g., `approx_distinct(...)` instead of `COUNT(DISTINCT ...)`
+  - See [approximate aggregation functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#approximate-aggregate-functions) in the standard SQL reference.
+- Reference the data size prediction ("This query will process X bytes") in STMO and the BigQuery UI to help gauge the efficiency of your queries. You should see this number go down as you limit the range of `submission_date`s or include fewer fields in your `SELECT` statement.
+- If using JOIN, trim the data to-be-joined before the query performs a JOIN. If you reduce data early in the processing cycle, shuffling and other complex operations only execute on the data that you need.
+  - Use sub queries with filters or intermediate tables or views as a way of decreasing sides of a join, prior to the join itself.
 - Do not treat WITH clauses as prepared statements
-    - WITH clauses are used primarily for readability because they are not materialized. For example, placing all your queries in WITH clauses and then running UNION ALL is a misuse of the WITH clause. If a query appears in more than one WITH clause, it executes in each clause.
-- Use approximate aggregation functions
-    - If the SQL aggregation function you're using has an equivalent approximation function, the approximation function will yield faster query performance. For example, instead of using `COUNT(DISTINCT)`, use `APPROX_COUNT_DISTINCT()`.
-    - See [approximate aggregation functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#approximate-aggregate-functions) in the standard SQL reference.
-- Reference the data size prediction ("This query will process X bytes") in STMO and the BigQuery UI to help gauge the efficiency of your queries. You should see this number go down as you limit the range of `submission_date`s or include fewer fields in your `SELECT` statement. For clustered tables, this estimate won't take into account benefits from `LIMIT`s and `WHERE` clauses on clustering fields, so you'll need to compare to the actual "Data Scanned" after the query is run. [Queries are charged by data scanned at $5/TB](https://cloud.google.com/bigquery/pricing#on_demand_pricing) so each 200 GB of data scanned will cost $1; it can be useful to keep the data estimate below 200 GB while developing and testing a query to limit cost and query time, then open up to the full range of data you need when you have confidence in the results.
+  - WITH clauses are used primarily for readability because they are not materialized. For example, placing all your queries in WITH clauses and then running UNION ALL is a misuse of the WITH clause. If a query appears in more than one WITH clause, it executes in each clause.
```

Comment:
This example of a misuse is confusing to me. In general it's fine to use CTEs any way users wish, just know they are not materialized.

Reply:
How about this?

```suggestion
  - WITH clauses are used primarily for readability because they are not materialized: if a query appears in more than one WITH clause, it executes in each clause. Do not rely on them to optimize your query!
```

Reply:
Perfect!

Reply:
Change fro:
If the SQL aggregation function you're using has an equivalent approximation function, the approximation function will yield faster query performance. 

To

If the SQL aggregation function that you use provides an equivalent approximation function, the approximation function then results in a much faster query performance. 
============

From

Reference the data size prediction ("This query will process X bytes") in STMO and the BigQuery UI to help gauge the efficiency of your queries. You should see this number go down as you limit the range of `submission_date`s or include fewer fields in your `SELECT` statement. For clustered tables, this estimate won't take into account benefits from `LIMIT`s and `WHERE` clauses on clustering fields, so you'll need to compare to the actual "Data Scanned" after the query is run. [Queries are charged by data scanned at $5/TB](https://cloud.google.com/bigquery/pricing#on_demand_pricing) so each 200 GB of data scanned will cost $1; it can be useful to keep the data estimate below 200 GB while developing and testing a query to limit cost and query time, then open up to the full range of data you need when you have confidence in the results.

To:


Reference the data size prediction ("This query processes X bytes") in STMO and the BigQuery UI to help evaluate the efficiency of your queries. You can view this number being decreased as you limit the range of `submission_date`s or include fewer fields in the `SELECT` statement. For clustered tables, this estimate does not take into account benefits from `LIMIT`s and `WHERE` clauses on clustering fields. Therefore, you need to compare to the actual "Data Scanned" after the query is run. [Queries are charged by data scanned at $5/TB](https://cloud.google.com/bigquery/pricing#on_demand_pricing). Each 200 GB of data scanned costs $1. You may want to keep the data estimate under 200 GB while you develop and test a query to limit cost and query time. You can then open it to the full range of data that you need when you have confidence in the results.